src/java.base/share/classes/java/lang/foreign/MemorySegment.java

```diff
@@ -176,28 +196,103 @@
- * matches the layout's {@linkplain MemoryLayout#byteAlignment() alignment constraints}. If the segment being
- * dereferenced is a native segment, then it has a concrete {@linkplain #address() base address}, which can
- * be used to perform the alignment check. The pseudo-function below demonstrates this:
+ * Access operations on a memory segment are constrained not only by the spatial and temporal bounds of the segment,
+ * but also by the <em>alignment constraint</em> of the value layout specified to the operation. An access operation can
+ * access only those offsets in the segment that denote addresses in physical memory which are <em>aligned</em> according
+ * to the layout. An address in physical memory is <em>aligned</em> according to a layout if the address is an integer
+ * multiple of the layout's alignment constraint. For example, the address 1000 is aligned according to an 8-byte alignment
+ * constraint (because 1000 is an integer multiple of 8), and to a 4-byte alignment constraint, and to a 2-byte alignment
+ * constraint; in contrast, the address 1004 is aligned according to a 4-byte alignment constraint, and to a 2-byte alignment
+ * constraint, but not to an 8-byte alignment constraint.
+ * Access operations are required to respect alignment because it can impact the performance of access operations, and
+ * can also determine which access operations are available at a given physical address. For instance,
+ * {@linkplain java.lang.invoke.VarHandle#compareAndSet(Object...) atomic access operations} operations using
+ * {@link java.lang.invoke.VarHandle} are only permitted at aligned addresses. In addition, alignment
+ * applies to an access operation whether the segment being accessed is a native segment or a heap segment.
+ * <p>
+ * If the segment being accessed is a native segment, then its {@linkplain #address() address} in physical memory can be
+ * combined with the offset to obtain the <em>target address</em> in physical memory. The pseudo-function below demonstrates this:
  *
- * {@snippet lang=java :
+ * {@snippet lang = java:
  * boolean isAligned(MemorySegment segment, long offset, MemoryLayout layout) {
- *   return ((segment.address().address() + offset) % layout.byteAlignment()) == 0;
+ *   return ((segment.address() + offset) % layout.byteAlignment()) == 0;
  * }
  * }
  *
- * If, however, the segment being dereferenced is a heap segment, the above function will not work: a heap
- * segment's {@linkplain #address() base address} is <em>virtualized</em> and, as such, cannot be used to construct an
- * alignment check. Instead, heap segments are assumed to produce addresses which are never more aligned than the element
- * size of the Java array from which they have originated from, as shown in the following table:
+ * For example:
+ * <ul>
+ * <li>A native segment at address 1000 can be accessed at offsets 0, 8, 16, 24, etc under an 8-byte alignment constraint,
```

Comment:
Here, we have to be more specific about the alignment of the segment address and the alignment of the layout otherwise there might be confusion.

Reply:
I think this includes all the information? E.g. address 1000 is a number, which is can be divided by 2/4/8.  In a paragraph above we explain that:

```
For example, the address 1000 is aligned according to an 8-byte alignment constraint (because 1000 is an integer multiple of 8), and to a 4-byte alignment constraint, and to a 2-byte alignment constraint; in contrast, the address 1004 is aligned according to a 4-byte alignment constraint, and to a 2-byte alignment constraint, but not to an 8-byte alignment constraint.
 ```

Reply:
I've decided to leave this as is for the time being.